slsim/Sources/QuasarCatalog/number_density_quasar.py

```diff
@@ -0,0 +1,246 @@
+        legend_handles_cdf.append(line)
+
+    return legend_handles_cdf
+
+
+def inverse_cdf_fits_for_redshifts(M_values, random_redshift_values):
+    """Creates inverse Cumulative Distribution Function (CDF) fits for each randomly
+    generated redshift.
+
+    Parameters:
+    :param M_values: Array of absolute i-band magnitudes.
+    :type M_values: numpy.ndarray
+
+    :param random_redshift_values: Array of randomly generated redshift values.
+    :type random_redshift_values: numpy.ndarray
+
+    Returns:
+    :rtype: dict
+    :return: Dictionary of inverse CDF functions for each redshift.
+    """
+
+    inverse_cdf_dict = {}
+
+    for random_redshift in random_redshift_values:
+        # Calculate cumulative probability
+        dPhi_dM_values = [dPhi_dM(M, random_redshift) for M in M_values]
+        cumulative_prob = np.cumsum(dPhi_dM_values) / np.sum(dPhi_dM_values)
+
+        # Interpolate the inverse CDF function
+        inverse_cdf = interp1d(
+            cumulative_prob, np.sort(M_values), kind="linear", fill_value="extrapolate"
+        )
+
+        # Store the inverse CDF function in the dictionary
+        inverse_cdf_dict[random_redshift] = inverse_cdf
+
+    return inverse_cdf_dict
+
+
+def generate_redshift_table(random_redshift_values, inverse_cdf_dict, seed=42):
+    """Generates random redshift values, associated inverse CDF values, and M values.
+
+    Parameters:
+    :param random_redshift_values: Array of randomly generated redshift values.
+    :type random_redshift_values: list or numpy.ndarray
+
+    :param inverse_cdf_dict: Dictionary of inverse CDF functions for each redshift.
+    :type inverse_cdf_dict: dict
+
+    :param seed: Seed for reproducibility (default: 42).
+    :type seed: int
+
+    Returns:
+    :rtype: astropy.table.Table
+    :return: Astropy table containing the generated data (Redshift, Inverse CDF Value, Associated M).
+    """
+    # Set the seed for reproducibility
+    np.random.seed(seed)
+
+    # Create an empty list to collect data for the table
+    table_data = {"Redshift": [], "Inverse_CDF_Value": [], "Associated_M": []}
+
+    # Generate random Inverse CDF value and associated M value for each redshift
+    for random_redshift in random_redshift_values:
+        # Find the inverse CDF function for the current random redshift
+        inverse_cdf = inverse_cdf_dict[random_redshift]
+
+        # Generate a random Inverse CDF value between 0 and 1
+        random_inverse_cdf_value = np.random.rand()
+
+        # Get the associated M value using the inverse CDF function
+        random_M_value = inverse_cdf(random_inverse_cdf_value)
+
+        # Append the data to the table
+        table_data["Redshift"].append(random_redshift)
+        table_data["Inverse_CDF_Value"].append(random_inverse_cdf_value)
+        table_data["Associated_M"].append(random_M_value)
+
+    # Create the Astropy table
+    table = Table(table_data)
+
+    return table
+
+
+# Overall test code usage
```

Comment:
please remove these lines of code from the commit file. If you want, you can have part of it in the test file or a notebook, but not necessary for the code commit